File: frontend/src/api/tenants.js
import { api } from "./client";

export async function listTenants() {
  const { data } = await api.get("/tenants");
  return data;
}

export async function createTenant(payload) {
  const { data } = await api.post("/tenants", payload);
  return data;
}

export async function createTenantUser(tenantId, payload) {
  const { data } = await api.post(`/tenants/${tenantId}/users`, payload);
  return data;
}

export async function establishTenantAccess(tenantId, reason) {
  const { data } = await api.post("/auth/platform/tenant-access", {
    tenant_id: tenantId,
    reason,
  });
  return data;
}

export async function tenantContext() {
  const { data } = await api.get("/tenant/context");
  return data;
}

File: backend/app/modules/tenants/routes/v1/tenants.py
from uuid import UUID
from fastapi import APIRouter, Depends
from app.core.dependencies import Principal, require_platform_permission
from app.modules.auth.schemas import CreateTenantRequest, TenantResponse
from app.modules.tenants.schemas import CreateTenantUserRequest, TenantUserResponse
from app.modules.tenants.service import create_tenant, list_tenants
from app.modules.tenants.users import create_tenant_user

router = APIRouter(prefix="/tenants", tags=["Tenants"])

@router.post("", response_model=TenantResponse, status_code=201)
async def create(payload: CreateTenantRequest, _: Principal = Depends(require_platform_permission("tenants.manage"))):
    return await create_tenant(payload)

@router.get("", response_model=list[TenantResponse])
async def list_all(_: Principal = Depends(require_platform_permission("tenants.manage"))):
    return await list_tenants()

@router.post("/{tenant_id}/users", response_model=TenantUserResponse, status_code=201)
async def create_user(tenant_id: UUID, payload: CreateTenantUserRequest, _: Principal = Depends(require_platform_permission("tenants.manage"))):
    return await create_tenant_user(tenant_id, payload)
